### PR TITLE
Benji/roomvc

### PR DESCRIPTION
**Pull request by Benjamindodgson**

### CODE REVIEW

File: Notification Content/NotificationViewController.swift [hunk 1]
```diff
@@ -16,14 +16,6 @@ class NotificationViewController: UIViewController, UNNotificationContentExtensi
 
     private var cancellables = Set<AnyCancellable>()
 
-    lazy var connectionRequestView: ConnectionRequestView = {
```
Comment (Benjamindodgson) [inline Notification Content/NotificationViewController.swift:19]: Not used

File: Shared/API/Connection+Cloud.swift [hunk 2]
```diff
@@ -20,7 +20,7 @@ struct CreateConnection: CloudFunction {
                      viewsToIgnore: [UIView]) async throws -> Connection {
         
         let params = ["to": self.to.objectId!,
-                      "status": Connection.Status.accepted.rawValue]
+                      "status": Connection.Status.invited.rawValue]
```
Comment (Benjamindodgson) [inline Shared/API/Connection+Cloud.swift:23]: Now we can handle connection requests with notices.

File: Shared/Coordinator/MainCoordinator+Launch.swift [hunk 3]
```diff
@@ -19,19 +19,12 @@ extension MainCoordinator {
             try? await ChatClient.initialize(for: User.current()!)
         }
         
-        let startingCID = deepLink?.conversationId
-        let startingMessageId = deepLink?.messageId
-        
-        if let coordinator = self.childCoordinator as? ConversationListCoordinator,
+        if let coordinator = self.childCoordinator as? RoomCoordinator,
```
Comment (Benjamindodgson) [inline Shared/Coordinator/MainCoordinator+Launch.swift:22]: Making the room coordinator the base.

File: Shared/Models/SystemAvatar.swift [hunk 4]
```diff
@@ -23,4 +23,8 @@ struct SystemAvatar: PersonType, Hashable {
     var phoneNumber: String?
 
     var image: UIImage?
+    
+    var updatedAt: Date? {
```
Comment (Benjamindodgson) [inline Shared/Models/SystemAvatar.swift:27]: Added so we can sort

File: Shared/Stores/NoticeStore.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+import Foundation
+import Combine
+import ParseLiveQuery
+import Parse
+import Localization
+
+class NoticeStore {
+
+    static let shared = NoticeStore()
```
Comment (Benjamindodgson) [inline Shared/Stores/NoticeStore.swift:17]: Very basic notice store similar to the people store.

File: Shared/Stores/PeopleStore.swift [hunk 6]
```diff
@@ -22,6 +22,7 @@ class PeopleStore {
     // MARK: - Public Events
     @Published var personUpdated: PersonType?
     @Published var personDeleted: PersonType?
+    @Published var personAdded: PersonType?
```
Comment (Benjamindodgson) [inline Shared/Stores/PeopleStore.swift:25]: Added so we can update the room members when someone new is added.

File: Shared/UI/Views/BorderedPersonView.swift [hunk 7]
```diff
@@ -12,7 +12,7 @@ class BorderedPersonView: PersonView {
     
     lazy var shadowLayer: CAShapeLayer = {
         let layer = CAShapeLayer()
-        layer.shadowColor = ThemeColor.yellow.color.cgColor
+        layer.shadowColor = ThemeColor.B6.color.cgColor
```
Comment (Benjamindodgson) [inline Shared/UI/Views/BorderedPersonView.swift:15]: Getting rid of empty yellow color.

File: iOS/Extensions/CNContact+Extensions.swift [hunk 8]
```diff
@@ -103,6 +103,10 @@ extension CNContact: PersonType {
     var image: UIImage? {
         return self.imageWith(text: self.initials)
     }
+    
+    var updatedAt: Date? {
+        return Date.distantPast
```
Comment (Benjamindodgson) [inline iOS/Extensions/CNContact+Extensions.swift:108]: Ensures these always show up at the end.

File: iOS/Flows/Conversation/Conversation List/ConversationListCoordinator.swift [hunk 9]
```diff
@@ -116,26 +112,6 @@ class ConversationListCoordinator: InputHandlerCoordinator<Void> {
     }
 }
 
-// MARK: - Permissions Flow
-
-extension ConversationListCoordinator {
-
-    @MainActor
-    func checkForPermissions() async {
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Conversation List/ConversationListCoordinator.swift:124]: Moved

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 10]
```diff
@@ -200,14 +203,17 @@ class ConversationListViewController: InputHandlerViewContoller, ConversationLis
...
-        guard let messageId = messageId else { return }
-
+        
         guard let cell = self.collectionView.cellForItem(at: conversationIndexPath),
               let messagesCell = cell as? ConversationMessagesCell else {
                   return
               }
 
+        guard let messageId = messageId else {
+            self.collectionView.scrollToItem(at: conversationIndexPath, at: .centeredHorizontally, animated: false)
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:213]: We may just want to scroll to the conversation, so this fixes that.

File: iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift [hunk 11]
```diff
@@ -220,7 +220,7 @@ class InputHandlerCoordinator<Result>: PresentableCoordinator<Result>,
 
 // MARK: - Image View Flow
 
-extension ConversationListCoordinator {
+extension InputHandlerCoordinator {
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift:223]: Not sure why this wasn't already this coordinator.

File: iOS/Flows/People/PeopleCoordinator.swift [hunk 12]
```diff
@@ -15,7 +15,7 @@ import Localization
 
 class PeopleCoordinator: PresentableCoordinator<[Person]> {
 
-    private lazy var peopleNavController = PeopleNavigationController()
+    private lazy var peopleNavController = PeopleNavigationController(showConnections: self.selectedConversationCID.exists)
```
Comment (Benjamindodgson) [inline iOS/Flows/People/PeopleCoordinator.swift:18]: No reason to show connections when you are just inviting people without a conversation. I tried thinking through the UX of inviting people and then adding them to conversations, but it just didn't make sense. Nor did it make sense to create one, then have everyone you invited be added to it. Maybe there is a better way...

File: iOS/Flows/Room/NoticeCell.swift [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+//  Jibber
+//
+//  Created by Benji Dodgson on 4/1/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+import StreamChat
+
+class NoticeCell: CollectionViewManagerCell, ManageableCell {
```
Comment (Benjamindodgson) [inline iOS/Flows/Room/NoticeCell.swift:12]: Basic Notice cell that can display each type.

File: iOS/Flows/Room/RoomCollectionViewDataSource.swift [hunk 14]
```diff
@@ -54,6 +81,24 @@ class RoomCollectionViewDataSource: CollectionViewDataSource<RoomSectionType, Ro
...
         switch section {
+        case .members:
+            if kind == UICollectionView.elementKindSectionHeader {
+                let header = collectionView.dequeueConfiguredReusableSupplementary(using: self.memberHeaderConfig, for: indexPath)
+                header.leftLabel.setText("Favorites")
+                header.button.didSelect { [unowned self] in
+                    self.didSelectAddPerson?()
+                }
+                return header
+            } else if kind == UICollectionView.elementKindSectionFooter {
```
Comment (Benjamindodgson) [inline iOS/Flows/Room/RoomCollectionViewDataSource.swift:92]: Added a footer and header to the members section as the conversations already has a  segment header, and a background that would be extended. This give the illusion of header for the conversations.

File: iOS/Flows/Room/RoomCoordinator+Notices.swift [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+//
+
+import Foundation
+import StreamChat
+
+extension RoomCoordinator {
+    
+    // The primary action
+    func handleRightOption(with notice: SystemNotice) {
+        switch notice.type {
```
Comment (Benjamindodgson) [inline iOS/Flows/Room/RoomCoordinator+Notices.swift:16]: Where we handle all interactions on a notice. Can only have 2 options. Most only have 1

File: iOS/Flows/Room/RoomCoordinator+Notices.swift [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+                    let text = "Your are now connected to \(user.fullName)."
+                    
+                    await ToastScheduler.shared.schedule(toastType: .basic(identifier: connectionId,
+                                                                     displayable: user,
+                                                                     title: "Connection Accepted",
+                                                                     description: text,
+                                                                     deepLink: nil))
+                    
+                    if let n = notice.notice {
+                        try n.delete()
```
Comment (Benjamindodgson) [inline iOS/Flows/Room/RoomCoordinator+Notices.swift:42]: Ideally we just listen to deletes from the server but for now, we can do this.

**Review by MartinJibber — APPROVED** [review]

**Pull request merged**